server/mqtt_ex_test_test.go

```diff
@@ -0,0 +1,569 @@
+			// configurations with different JetStream domains. Not sure if it
+			// is a bug.
+
+			// initialize the MQTT assets by "touching" all nodes in the
+			// cluster, but then reload to start with fresh server state.
+			for _, dial := range target.all {
+				mqttExInitServer(t, dial)
+			}
+			target.Reload(t)
+
+			numRMS := 100
+			strNumRMS := strconv.Itoa(numRMS)
+			for _, tc := range target.configs {
+				t.Run(tc.name, func(t *testing.T) {
+					topic := "subret_" + nuid.Next()
+
+					// publish numRMS retained messages, one at a time,
+					// round-robin across pub nodes.
+					iNode := 0
+					for i := 0; i < numRMS; i++ {
+						pubTopic := fmt.Sprintf("%s/%d", topic, i)
+						dial := tc.pub[iNode%len(tc.pub)]
+						mqttexRunTest(t, "pub", dial,
+							"--retain",
+							"--topic", pubTopic,
+							"--qos", "0",
+							"--size", "100",
+						)
+						iNode++
+					}
+
+					check := func(testNameFormat string) {
+						for _, dial := range tc.sub {
+							_, _, _, name := dial.Get()
+							t.Run(fmt.Sprintf(testNameFormat, name), func(t *testing.T) {
+								mqttexRunTest(t, "sub", dial,
+									"--retained", strNumRMS,
+									"--qos", "0",
+									"--topic", topic,
+								)
+							})
+						}
+					}
+
+					// Subscribe and receive all retained on each sub node. Then
+					// reload the entire target cluster, and repeat to ensure
+					// retained messages are persisted and re-delivered after
+					// restart.
+					check("subscribe at %s")
+					target.Reload(t)
+					check("subscribe after reload at %s")
+				})
+			}
+		})
+	}
+}
+
+func mqttExInitServer(tb testing.TB, dial mqttExDial) {
+	tb.Helper()
+	mqttexRunTest(tb, "pub", dial, "--id", "__init__")
+}
+
+func mqttExNewDialForServer(s *Server, username, password string) mqttExDial {
+	o := s.getOpts().MQTT
+	return mqttExNewDial(username, password, o.Host, o.Port, s.Name())
+}
+
+func mqttExNewDial(username, password, host string, port int, comment string) mqttExDial {
+	d := ""
+	switch {
+	case username != "" && password != "":
+		d = fmt.Sprintf("%s:%s@%s:%d", username, password, host, port)
+	case username != "":
+		d = fmt.Sprintf("%s@%s:%d", username, host, port)
+	default:
+		d = fmt.Sprintf("%s:%d", host, port)
+	}
+	if comment != "" {
+		d += "#" + comment
+	}
+	return mqttExDial(d)
+}
+
+func (d mqttExDial) Get() (u, p, s, c string) {
+	if d == "" {
+		return "", "", "127.0.0.1:1883", ""
+	}
+	in := string(d)
+	if i := strings.LastIndex(in, "#"); i != -1 {
+		c = in[i+1:]
+		in = in[:i]
+	}
+	if i := strings.LastIndex(in, "@"); i != -1 {
+		up := in[:i]
+		in = in[i+1:]
+		u = up
+		if i := strings.Index(up, ":"); i != -1 {
+			u = up[:i]
+			p = up[i+1:]
+		}
+	}
+	s = in
+	return u, p, s, c
+}
+
+func (t *mqttExTarget) Reload(tb testing.TB) {
+	tb.Helper()
+
+	for _, c := range t.clusters {
+		c.restartAllSamePorts()
+	}
+
+	for i, s := range t.servers {
+		o := s.getOpts()
+		s.Shutdown()
```

Comment:
It's the case at the moment, but you're right that it may not be the case in the future (a cluster with a single leaf node immediately comes to mind).

TBH, this was just a "minimal viable" implementation for the current requirements, I didn't want to cut/paste nor to go too abstract here (it's just an "extra" test for the time being). If we keep using MQTTEx going forward more then yes, the "target" implementation should be refactored.